packages/gatsby/cache-dir/dev-loader.js

```diff
@@ -138,6 +143,41 @@ class DevLoader extends BaseLoader {
       ___emitter.emit(`pageQueryResult`, newPageData)
     }
   }
+
+  handleDirtyPageQueryMessage(msg) {
+    msg.payload.dirtyQueries.forEach(dirtyQueryId => {
+      if (dirtyQueryId === `/dev-404-page/` || dirtyQueryId === `/404.html`) {
+        // those pages are not on demand so skipping
+        return
+      }
+
+      const normalizedId = normalizePagePath(dirtyQueryId)
+
+      // We can't just delete items in caches, because then
+      // using history.back() would show dev-404 page
+      // due to our special handling of it in root.js (loader.isPageNotFound check)
+      // so instead we mark it as stale and instruct loader's async methods
+      // to refetch resources if they are marked as stale
+
+      const cachedPageData = this.pageDataDb.get(normalizedId)
+      if (cachedPageData) {
+        // if we have page data in cache, mark it as stale
+        this.pageDataDb.set(normalizedId, {
+          ...cachedPageData,
+          stale: true,
+        })
+      }
+
+      const cachedPage = this.pageDb.get(normalizedId)
+      if (cachedPage) {
+        // if we have page data in cache, mark it as stale
+        this.pageDb.set(normalizedId, {
+          ...cachedPage,
+          payload: { ...cachedPage.payload, stale: true },
```

Comment:
Is there a reason to set `stale` flag in the `payload` for `cachedPage` case here but not for the `pageDataDb` case above (there it's just in the root)?

Can we have both at the root level for consistency?

Reply:
I would love this to be more consistent but I did this because
https://github.com/gatsbyjs/gatsby/pull/28127/commits/02c2ea8053b4501036e7f90129464380c26990a7#diff-2849c56428d0997b1a7c184b6b1348dd709c302cff236977951468666d85fa0fR50-R57

Only get `payload` part of page entry (from `pageDb`), so I would need to refactor quite a bit more of runtime to be able to access it there

Reply:
Oh well, that makes sense. Will have to stick with what we have for now.